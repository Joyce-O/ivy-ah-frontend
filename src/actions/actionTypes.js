export const AUTH_LOADING = 'AUTH_LOADING';
export const LOGIN = 'LOGIN';
export const TOGGLE_MODAL = 'TOGGLE_MODAL';
export const FETCH_ARTICLE_STARTED = 'FETCH_ARTICLE_STARTED';
export const FETCH_ARTICLE_SUCCEEDED = 'FETCH_ARTICLE_SUCCEEDED';
export const FETCH_ARTICLE_FAILED = 'FETCH_ARTICLE_FAILED';
export const EDIT_BOOKMARKS_STARTED = 'EDIT_BOOKMARKS_STARTED';
export const FETCH_BOOKMARKS_SUCCEEDED = 'FETCH_BOOKMARKS_SUCCEEDED';
export const ADD_BOOKMARKS_SUCCEEDED = 'ADD_BOOKMARKS_SUCCEEDED';
export const REMOVE_BOOKMARKS_SUCCEEDED = 'REMOVE_BOOKMARKS_SUCCEEDED';
export const EDIT_BOOKMARKS_FAILED = 'EDIT_BOOKMARK_FAILED';
export const AUTH_SUCCESS = 'AUTH_SUCCESS';
export const AUTH_FAIL = 'AUTH_FAIL';
export const VERIFY_EMAIL = 'VERIFY_EMAIL';
