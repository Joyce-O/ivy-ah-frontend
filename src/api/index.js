import axios from 'axios';

const API_BASE_URL = 'https://ivy-ah-backend-staging.herokuapp.com/api/v1';

const token = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6Iml2eWFkbWluQGdtYWlsLmNvbSIsImlkIjoiMDU1MDk2YWUtMDcyYy00MjJkLTgwMTctZDM2ZTFiZWNmNTNjIiwicm9sZSI6ImFkbWluIiwiaWF0IjoxNTU3MTUzNzc2fQ.SeUSBXKom7eHkEXgak8uS-3xqVKSwR3DvrGyIcn7CSs';

export const client = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
    Authorization: token
  },
});
export const signup = params => client.post('/users/signup', params);
export const login = params => client.post('/users/login', params);
export const fetchBookmarks = () => client.get('/bookmarks');
export const addBookmark = article => client.post('/bookmarks', { article });
export const removeBookmark = article => client.delete('/bookmarks', {
  data: {
    article
  }
});
export const fetchArticle = id => client.get(`/articles/${id}`);
