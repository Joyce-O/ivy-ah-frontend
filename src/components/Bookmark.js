import React from 'react';
import PropTypes from 'prop-types';

const Bookmark = ({ bookmark, onClickBookmark }) => (
  <button type="button" className="bookmark-button" onClick={() => onClickBookmark(!bookmark)}>
    <i className={bookmark ? 'bookmark icon' : 'bookmark outline icon'} />
  </button>
);

Bookmark.propTypes = {
  bookmark: PropTypes.bool.isRequired,
  onClickBookmark: PropTypes.func.isRequired,
};

export default Bookmark;
