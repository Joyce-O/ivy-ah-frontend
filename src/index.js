import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import 'semantic-ui-css/semantic.min.css';
import 'overlayscrollbars/css/OverlayScrollbars.min.css';
import 'react-draft-wysiwyg/dist/react-draft-wysiwyg.css';
import App from './containers/App';
import Store from './store';
import './assets/sass/index.scss';

render(<Provider store={Store}><App /></Provider>, document.getElementById('root'));
