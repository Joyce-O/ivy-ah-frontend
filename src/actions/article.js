import * as api from '../api';
import * as actions from './actionTypes';

export const fetchArticlesStarted = () => ({
  type: actions.FETCH_ALL_ARTICLES_STARTED,
});

const fetchArticlesSucceeded = articles => ({
  type: actions.FETCH_ALL_ARTICLES_SUCCEEDED,
  payload: { articles }
});

const fetchArticlesFailed = err => ({
  type: actions.FETCH_ALL_ARTICLES_FAILED,
  payload: { err }
});

export const fetchArticles = (page, limit) => (dispatch) => {
  dispatch(fetchArticlesStarted());
  return api.fetchArticles(page, limit).then((res) => {
    dispatch(fetchArticlesSucceeded(res.data));
  }).catch((err) => {
    dispatch(fetchArticlesFailed(err.message));
  });
};
