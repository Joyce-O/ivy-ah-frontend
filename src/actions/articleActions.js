import * as api from '../api';
import * as actions from './actionTypes';

export const fetchArticleStarted = () => ({
  type: actions.FETCH_ARTICLE_STARTED
});

export const fetchArticleSucceeded = article => ({
  type: actions.FETCH_ARTICLE_SUCCEEDED,
  payload: { article }
});

export const fetchArticleFailed = error => ({
  type: actions.FETCH_ARTICLE_FAILED,
  payload: { error }
});

export const fetchArticle = id => (dispatch) => {
  dispatch(fetchArticleStarted());
  return api.fetchArticle(id)
    .then((response) => {
      dispatch(fetchArticleSucceeded(response.data));
    })
    .catch(error => dispatch(fetchArticleFailed(error)));
};
