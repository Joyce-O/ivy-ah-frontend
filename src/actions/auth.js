import * as actions from './actionTypes';
import { signup, login } from '../api';


export const authStart = () => ({
  type: actions.AUTH_LOADING
});

export const toggleModal = (modalPane = 'sign in') => ({
  type: actions.TOGGLE_MODAL,
  modalPane
});

export const authSuccess = (token, id, email) => ({
  type: actions.AUTH_SUCCESS,
  userId: id,
  email,
  token
});

export const authFail = error => ({
  type: actions.AUTH_FAIL,
  error
});

export const verifyEmail = email => ({
  type: actions.VERIFY_EMAIL,
  email
});

export const signUp = (firstname, lastname, email, password) => (dispatch) => {
  dispatch(authStart());
  const authData = {
    firstname,
    lastname,
    email,
    password,
  };
  signup(authData)
    .then((response) => {
      dispatch(verifyEmail(response.data.user.email));
    })
    .catch((err) => {
      const { error } = err.response.data;
      dispatch(authFail(error));
    });
};

export const logIn = (email, password) => (dispatch) => {
  dispatch(authStart());
  const authData = {
    email,
    password,
  };
  login(authData)
    .then((response) => {
      const { userid, token } = response.data.user;
      if (!response.data.user.isVerified) {
        dispatch(verifyEmail(response.data.user.email));
      } else {
        localStorage.setItem('user', JSON.stringify(response.data.user));
        dispatch(authSuccess(token, userid, response.data.user.email));
      }
    })
    .catch((err) => {
      const { error } = err.response.data;
      dispatch(authFail(error));
    });
};
