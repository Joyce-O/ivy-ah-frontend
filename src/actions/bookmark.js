import * as api from '../api';
import * as actions from './actionTypes';


export const editBookmarkStarted = (articleId, bookmarked) => ({
  type: actions.EDIT_BOOKMARKS_STARTED,
  payload: { articleId, bookmarked }
});

export const editBookmarksFailed = articleId => ({
  type: actions.EDIT_BOOKMARKS_FAILED,
  payload: articleId
});

export const fetchBookmarksSucceeded = bookmarks => ({
  type: actions.FETCH_BOOKMARKS_SUCCEEDED,
  payload: { bookmarks }
});

export const fetchBookmarks = () => dispatch => api.fetchBookmarks().then((res) => {
  dispatch(fetchBookmarksSucceeded(res.data));
});

export const addBookmarksSucceeded = bookmarks => ({
  type: actions.ADD_BOOKMARKS_SUCCEEDED,
  payload: { bookmarks }
});

export const addBookmarks = articleId => (dispatch) => {
  dispatch(editBookmarkStarted(articleId, true));
  return api.addBookmark(articleId).then((res) => {
    dispatch(addBookmarksSucceeded(res.data.bookmarks));
  }).catch(() => {
    dispatch(editBookmarksFailed(articleId));
  });
};

export const removeBookmarksSucceeded = bookmarks => ({
  type: actions.REMOVE_BOOKMARKS_SUCCEEDED,
  payload: { bookmarks }
});

export const removeBookmarks = articleId => (dispatch) => {
  dispatch(editBookmarkStarted(articleId, false));
  return api.removeBookmark(articleId).then(() => {
    dispatch(removeBookmarksSucceeded(articleId));
  }).catch(() => {
    dispatch(editBookmarksFailed(articleId));
  });
};
