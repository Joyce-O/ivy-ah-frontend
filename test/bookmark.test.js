import React from 'react';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import bookmarksReducer from '../src/reducers/bookmark';
import Bookmark from '../src/components/Bookmark';
import * as api from '../src/api';
import * as actions from '../src/actions/actionTypes';
import {
  fetchBookmarksSucceeded,
  fetchBookmarks,
  addBookmarks,
  removeBookmarks
} from '../src/actions/bookmark';

const articleId = 'f2fba6df-75d5-4672-9923-834a6da4e11e';
const onClickBookmark = jest.fn();
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

it("renders an empty icon if article isn't bookmarked", () => {
  const bookmark = false;
  const tree = renderer
    .create(<Bookmark
      articleId={articleId}
      bookmark={bookmark}
      onClickBookmark={onClickBookmark}
    />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders a filled icon if article has been bookmarked', () => {
  const bookmark = true;
  const tree = renderer
    .create(<Bookmark
      articleId={articleId}
      bookmark={bookmark}
      onClickBookmark={onClickBookmark}
    />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});


describe('addBookmark', () => {
  api.client.post = jest.fn(() => new Promise((resolve, reject) => resolve({
    data: { bookmarks: 'bookmark' }
  })));
  it('addBookmarks', () => {
    const getState = {};
    const expectedActions = {
      type: actions.EDIT_BOOKMARKS_STARTED,
      payload: {
        articleId: 'bookmark',
        bookmarked: true
      },
    };

    const store = mockStore(getState, expectedActions);
    return store.dispatch(addBookmarks('bookmark')).then(() => {
      expect(store.getActions()[0]).toEqual(expectedActions);
      expect(api.client.post).toHaveBeenCalled();
      expect(api.client.post).toHaveBeenCalledWith('/bookmarks', { article: 'bookmark' });
    });
  });

  it('addBookmarksFailed', () => {
    api.client.post = jest.fn(() => new Promise((resolve, reject) => reject(new Error())));

    const getState = {};
    const expectedAction = [
      {
        payload: { articleId: 'bookmark', bookmarked: true }, type: 'EDIT_BOOKMARKS_STARTED'
      },
      { payload: 'bookmark', type: 'EDIT_BOOKMARK_FAILED' }
    ];

    const store = mockStore(getState, expectedAction);
    return store.dispatch(addBookmarks('bookmark')).then(() => {
      expect(store.getActions()).toEqual(expectedAction);
      expect(api.client.post).toHaveBeenCalled();
    });
  });
});

describe('removeBookmark', () => {
  it('removeBookmarksSucceeded', () => {
    api.client.delete = jest.fn(() => new Promise((resolve, reject) => resolve({
      data: { bookmarks: 'bookmark' }
    })));
    const getState = {};
    const expectedActions = {
      type: actions.EDIT_BOOKMARKS_STARTED,
      payload: {
        articleId: 'bookmark',
        bookmarked: false
      },
    };

    const store = mockStore(getState, expectedActions);
    return store.dispatch(removeBookmarks('bookmark')).then(() => {
      expect(store.getActions()[0]).toEqual(expectedActions);
      expect(api.client.delete).toHaveBeenCalled();
      expect(api.client.delete).toHaveBeenCalledWith('/bookmarks', { data: { article: 'bookmark' } });
    });
  });
});

describe('fetchBookmark', () => {
  api.client.get = jest.fn(() => new Promise((resolve, reject) => resolve({
    data: [
      {
        article: 'cd1fc24e-0a1d-4a55-8798-1a6251a7a53e'
      },
      {
        article: 'b52da1ce-c138-4494-8163-b56f78ff7a71'
      }
    ]
  })));
  const bookmarks = [
    {
      article: 'cd1fc24e-0a1d-4a55-8798-1a6251a7a53e'
    },
    {
      article: 'b52da1ce-c138-4494-8163-b56f78ff7a71'
    },

  ];

  it('fetchBookmarksSucceeded', () => {
    const expectedAction = {
      type: actions.FETCH_BOOKMARKS_SUCCEEDED,
      payload: {
        bookmarks
      },
    };
    expect(fetchBookmarksSucceeded(bookmarks)).toEqual(expectedAction);
  });

  it('fetchBookmarks', () => {
    const getState = {};
    const expectedActions = {
      type: actions.FETCH_BOOKMARKS_SUCCEEDED,
      payload: { bookmarks },
    };

    const store = mockStore(getState, expectedActions);
    return store.dispatch(fetchBookmarks({})).then(() => {
      expect(store.getActions()[0]).toEqual(expectedActions);
      expect(api.client.get).toHaveBeenCalled();
      expect(api.client.get).toHaveBeenCalledWith('/bookmarks');
    });
  });
});

describe('the bookmarks reducer', () => {
  const initialState = {
    bookmarks: [],
    error: false,
    target: false,
    bookmarked: null
  };

  it('should return the intial state', () => {
    expect(bookmarksReducer(undefined, {})).toEqual(initialState);
  });

  it('should handle edit bookmarks started', () => {
    const action = {
      type: actions.EDIT_BOOKMARKS_STARTED,
      payload: {
        articleId: 'articleId',
        bookmarked: 'boolean'
      }
    };
    const expectedState = {
      ...initialState,
      target: action.payload.articleId,
      bookmarked: action.payload.bookmarked,
      error: false
    };

    expect(bookmarksReducer(initialState, action)).toEqual(expectedState);
  });

  it('should handle fetch bookmarks succeeded', () => {
    const action = {
      type: actions.FETCH_BOOKMARKS_SUCCEEDED,
      payload: {
        bookmarks: {
          status: 'http status code',
          bookmarks: ['bookmark items']
        }
      }
    };
    const expectedState = {
      ...initialState,
      bookmarks: action.payload.bookmarks.bookmarks
    };

    expect(bookmarksReducer(initialState, action)).toEqual(expectedState);
  });

  it('should handle add bookmarks succeeded', () => {
    const action = {
      type: actions.ADD_BOOKMARKS_SUCCEEDED,
      payload: {
        bookmarks: [
          { article: 'articleId' }
        ]
      }
    };
    const expectedState = {
      ...initialState,
      bookmarks: initialState.bookmarks.concat(action.payload.bookmarks),
      target: action.payload.bookmarks.article,
      bookmarked: null,
      error: false
    };

    expect(bookmarksReducer(initialState, action)).toEqual(expectedState);
  });

  it('should handle remove bookmarks succeeded', () => {
    const action = {
      type: actions.REMOVE_BOOKMARKS_SUCCEEDED,
      payload: {
        bookmarks: [
          { article: 'articleId' }
        ]
      }
    };
    const expectedState = {
      ...initialState,
      bookmarks: initialState.bookmarks
        .filter(bookmark => bookmark.article !== action.payload.bookmarks),
      target: action.payload.bookmarks,
      bookmarked: null,
      error: false
    };

    expect(bookmarksReducer(initialState, action)).toEqual(expectedState);
  });

  it('should handle edit bookmarks failed', () => {
    const action = {
      type: actions.EDIT_BOOKMARKS_FAILED,
      payload: {
        bookmarks: [
          { article: 'articleId' }
        ]
      }
    };
    const expectedState = {
      ...initialState,
      error: true,
      target: action.payload,
      bookmarked: null
    };

    expect(bookmarksReducer(initialState, action)).toEqual(expectedState);
  });
});
