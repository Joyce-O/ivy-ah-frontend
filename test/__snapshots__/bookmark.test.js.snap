// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a filled icon if article has been bookmarked 1`] = `
<button
  className="bookmark-button"
  onClick={[Function]}
  type="button"
>
  <i
    className="bookmark icon"
  />
</button>
`;

exports[`renders an empty icon if article isn't bookmarked 1`] = `
<button
  className="bookmark-button"
  onClick={[Function]}
  type="button"
>
  <i
    className="bookmark outline icon"
  />
</button>
`;
